tests/instruments/test_instrument.py

```diff
@@ -234,18 +258,89 @@ class Fake(FakeInstrument):
 
 
 @pytest.mark.parametrize(
-    'set_command, given, expected',
-    [("%d", 5, 5),
-     ("%d, %d", (5, 6), [5, 6]),  # input has to be a tuple, not a list
+    'set_command, given, expected, dynamic',
+    [("%d", 5, 5, False),
+     ("%d", 5, 5, True),
+     ("%d, %d", (5, 6), [5, 6], False),  # input has to be a tuple, not a list
+     ("%d, %d", (5, 6), [5, 6], True),  # input has to be a tuple, not a list
      ])
-def test_fakeinstrument_control(set_command, given, expected):
+def test_fakeinstrument_control(set_command, given, expected, dynamic):
     """FakeInstrument's custom simple control needs to process values correctly.
     """
     class Fake(FakeInstrument):
         x = FakeInstrument.control(
             "", set_command, "",
+            dynamic=dynamic,
         )
 
     fake = Fake()
     fake.x = given
     assert fake.x == expected
+
```

Comment:
I have reworked the dynamic property tests and split them up to have more fine-grained resolution if something breaks. What do you think? I could not implement this as a ready-to-merge Github "suggestion", so can only deliver this way.
Replace from here with:
```

class GenericInstrument(FakeInstrument):
    #  Use truncated_range as this easily lets us test for the range boundaries
    fake_ctrl = Instrument.control(
        "", "%d", "docs",
        validator=truncated_range,
        values=(1, 10),
        dynamic=True,
    )
    fake_setting = Instrument.setting(
        "%d", "docs",
        validator=truncated_range,
        values=(1, 10),
        dynamic=True,
    )
    fake_measurement = Instrument.measurement(
        "", "docs",
        values={'X': 1, 'Y': 2, 'Z': 3},
        map_values=True,
        dynamic=True,
    )


class ExtendedInstrument(GenericInstrument):
    # Keep values unchanged, just derive another instrument, e.g. to add more properties
    pass


class StrictExtendedInstrument(ExtendedInstrument):
    # Use strict instead of truncated range validator
    fake_ctrl_validator = strict_range
    fake_setting_validator = strict_range


class NewRangeInstrument(GenericInstrument):
    # Choose different properties' values, like you would for another device model
    fake_ctrl_values = (10, 20)
    fake_setting_values = (10, 20)
    fake_measurement_values = {'X': 4, 'Y': 5, 'Z': 6}


def test_dynamic_property_unchanged_by_inheritance():
    generic = GenericInstrument()
    extended = ExtendedInstrument()

    generic.fake_ctrl = 50
    assert generic.fake_ctrl == 10
    extended.fake_ctrl = 50
    assert extended.fake_ctrl == 10

    generic.fake_setting = 50
    assert generic.read() == '10'
    extended.fake_setting = 50
    assert extended.read() == '10'

    generic.read()
    generic.write('1')
    assert generic.fake_measurement == 'X'
    extended.read()
    extended.write('1')
    assert extended.fake_measurement == 'X'


def test_dynamic_property_strict_raises():
    strict = StrictExtendedInstrument()

    with pytest.raises(ValueError):
        strict.fake_ctrl = 50
    with pytest.raises(ValueError):
        strict.fake_setting = 50


def test_dynamic_property_values_update_in_subclass():
    newrange = NewRangeInstrument()

    newrange.fake_ctrl = 50
    assert newrange.fake_ctrl == 20

    newrange.fake_setting = 50
    assert newrange.read() == '20'

    newrange.read()
    newrange.write('4')
    assert newrange.fake_measurement == 'X'


def test_dynamic_property_values_update_in_instance():
    generic = GenericInstrument()

    generic.fake_ctrl_values = (0, 33)
    generic.fake_ctrl = 50
    assert generic.fake_ctrl == 33

    generic.fake_setting_values = (0, 33)
    generic.fake_setting = 50
    assert generic.read() == '33'

    generic.fake_measurement_values = {'X': 7}
    generic.read()
    generic.write('7')
    assert generic.fake_measurement == 'X'


def test_dynamic_property_values_update_in_one_instance_leaves_other_unchanged():
    generic1 = GenericInstrument()
    generic2 = GenericInstrument()

    generic1.fake_ctrl_values = (0, 33)
    generic1.fake_ctrl = 50
    generic2.fake_ctrl = 50
    assert generic1.fake_ctrl == 33
    assert generic2.fake_ctrl == 10


def test_dynamic_property_reading_special_attributes_forbidden():
    generic = GenericInstrument()
    with pytest.raises(AttributeError):
        generic.fake_ctrl_validator
```

Reply:
Ok, thanks for the code. I will check the feedback on the code part before updating the tests

Reply:
Ok, I will make the changes, testing seems more structured this way. Only one remark, I removed some dummy read here and there in the test code probably result of some test debug, I mean they do not harm, but I am not sure they are serving any purpose.